storage/src/mmr/hasher.rs

```diff
@@ -181,36 +183,96 @@ impl<'a, H: CHasher, S: Storage<H::Digest>> Grafting<'a, H, S> {
+///
+/// This algorithm performs walks down corresponding branches of the peak and base trees. When we
+/// find the node in the peak tree we are looking for, we return the position of the corresponding
+/// node reached in the base tree.
+fn destination_pos(peak_node_pos: u64, height: u32) -> u64 {
+    let leading_zeros = (peak_node_pos + 1).leading_zeros();
+    let mut peak_pos = u64::MAX >> leading_zeros;
+    let mut base_pos = u64::MAX >> (leading_zeros - height);
+    let mut peak_height = peak_pos.trailing_ones() - 1;
+    let mut base_height = peak_height + height;
+    peak_pos -= 1;
+    base_pos -= 1;
+
+    while base_height >= height {
+        if peak_pos == peak_node_pos {
+            break;
+        }
+
+        let left_pos = peak_pos - (1 << peak_height);
+        if left_pos < peak_node_pos {
+            peak_pos -= 1;
+            base_pos -= 1;
+        } else {
+            peak_pos = left_pos;
+            base_pos -= 1 << base_height;
+        }
+
+        peak_height -= 1;
+        base_height -= 1;
+    }
+
+    base_pos
+}
+
+/// Inverse computation of destination_pos, with an analogous implementation involving walks down
+/// corresponding branches of both trees. Returns none if there is no corresponding node.
+fn source_pos(base_node_pos: u64, height: u32) -> Option<u64> {
+    if pos_to_height(base_node_pos) < height {
+        // Nodes below the grafting height do not have a corresponding peak tree node.
+        return None;
+    }
 
-        // The rightmost-leaf in the corresponding segment of the peak tree:
-        let base_leaf_num = (peak_leaf_num * chunk_size_bits) + chunk_size_bits - 1;
+    let leading_zeros = (base_node_pos + 1).leading_zeros();
+    let mut base_pos = u64::MAX >> leading_zeros;
+    let mut peak_pos = u64::MAX >> (leading_zeros + height);
+    let mut base_height = base_pos.trailing_ones() - 1;
+    let mut peak_height = base_height - height;
+    base_pos -= 1;
+    peak_pos -= 1;
+
+    while base_pos != base_node_pos {
+        let left_pos = base_pos - (1 << base_height);
+        if left_pos < base_node_pos {
+            base_pos -= 1;
+            peak_pos -= 1;
+        } else {
+            base_pos = left_pos;
+            peak_pos -= 1 << peak_height;
+        }
 
-        // Walking up self.height levels from the rightmost leaf involves simply adding self.height
-        // to its position.
-        leaf_num_to_pos(base_leaf_num) + self.height as u64
+        base_height -= 1;
+        peak_height -= 1;
     }
+
+    Some(peak_pos)
 }
 
 impl<H: CHasher, S: Storage<H::Digest>> Hasher<H> for Grafting<'_, H, S> {
     async fn leaf_digest(&mut self, pos: u64, element: &[u8]) -> Result<H::Digest, Error> {
         let base_node_pos = self.destination_pos(pos);
         if base_node_pos >= self.base_mmr.size() {
             // The base tree doesn't yet have a node where we can graft this position.
+            // TODO: change this to assert-fail on this condition once
```

Comment:
This is no merged 🫡

Reply:
Yup, rebased & this TODO is now addressed.